icefoganalytics/travel-authorization
# Basic Estimates Tab - Part 1 - Estimates CRUD Table

References
- [Modeling, UI Label To Data Model Mapping for Travel Authorization Form, 2023-10-05](https://docs.google.com/document/d/1SXCUgpUMgFLESLaHZa-Duv03zzJbRDZ2BJCxmN7GXao/edit?pli=1)

Fixes https://github.com/icefoganalytics/travel-authorization/issues/11

# User Story

As a User,
I want the system to create an estimate of my trip,
So that I can plan my journey and budget accordingly.

# Context

On the `/my-travel-request/create` page, there should be a tab for "Estimates".
There should also be the edit version of this page at `/my-travel-requests/:form-id`.
These tabs should be accessible via url, e.g. `/my-travel-requests/create/estimates` or something.

![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/f76ea01a-b6a7-42c9-b7e3-ded8831d490f)

# Implementation

1. Switch from a travel authorization request "create", then "edit" interface to:
    1. On click of "+ Travel Authorization" creation button, immediately creating a blank "draft" request
    2. Redirecting the user to the edit page for the newly create authorization.
    
    This provides several advantages, the primary one being that the user can save the form at any point.
2. Built out a basic create/read/update/delete (CRUD) interface for Estimate creation. The next stage of this project will be to initially replace the "Add Estimate" button with a "Generate Estimates" button. The "generate" button should pull generate from the form "stops". After the estimates have been generated, the user will then be able to edit the estimates at will.

# Screenshots

`/my-travel-requests/19/estimate`
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/38c812ac-1be4-43a4-880e-7051f7f4680e)

`/my-travel-requests/19/details`
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/ad6b090f-bab2-4919-b5c6-93160917a0a7)


# Testing Instructions

1. Boot the app database via `dev up`.
4. Boot the app api service via `cd src/api && npm run start`.
5. Boot the app web service via `cd src/web && npm run start`.
6. Check that the app boots correctly and that you can log in.
7. Go to "My Travel Requests" from the top navigation drop down.
8. In the table, click on a "travel request" with a "Submitted" status.
9. You will be redirected to the "details" tab of the request.
10. Click on the "Estimate" tab to view estimates for that request.
11. Check that you get an error message when you try to create/edit/delete any estimates.
12. Go back to the "My Travel Requests" page via the breadcrumbs.
13. In the table, click on a "travel request" with a "Draft" status, or create a new "Travel Authorization" if you don't have any.
14. Click on the "Estimate" tab to view estimates for that request.
15. Check that you can create/update/delete estimates.
16. Check that when you have a create/update/delete modal active that the URL updates.
17. Check that the modal is auto-opened if you navigate directly to the url or refresh the page.
18. Go back to the "Details" tab, and scroll down until you reach the "Approvals" section.
19. Check that the "Estimated Cost" field matches the computed total from the "Estimate" tab.

# Deploy Instructions

1. Run migrations via going to the `/api/migrate/latest` endpoint after deployment.
